src/merkle.rs

```diff
@@ -0,0 +1,332 @@
+        1 << (self.levels - 1)
+    }
+
+    // A Merkle Tree can be built out of an array.
+    pub fn from_array<T: Hash, const count: usize>(array: [T; count]) -> Self {
+        let mut levels: usize = 1;
+        let mut hashes: Vec<Vec<u64>> = vec![Vec::new()];
+
+        for element in array {
+            hashes[0].push(hash(element, None));
+        }
+
+        while hashes[levels - 1].len() > 1 {
+            hashes.push(Vec::new());
+
+            let mut index = 0;
+
+            while index < hashes[levels - 1].len() {
+                let left = hashes[levels - 1][index];
+                let right = match hashes[levels - 1].get(index + 1) {
+                    Some(value) => *value,
+                    None => left,
+                };
+                hashes[levels].push(hash(left, Some(right)));
+                index += 2;
+            }
+
+            levels += 1;
+        }
+
+        MerkleTree {
+            hashes,
+            levels,
+            count,
+        }
+    }
+
+    // A Merkle Tree can be dynamic, this means that elements can be added once it is built.
+    //Inserts an element and returns the index where it has been inserted
+    pub fn insert_element<T: Hash>(&mut self, element: T) -> usize {
+        let mut index = self.count;
+        self.count += 1;
+
+        // If we're at the limit of size, we need a new level to combine the new leaf with the rest
+        if self.count > self.max_count() {
+            self.hashes.push(vec![0]);
+            self.levels += 1;
+        }
+
+        // We update the affected parent for each level, which is always the rightmost one
+        let mut new_hash = hash(element, None);
+        self.hashes[0].push(new_hash);
+        for i in 1..self.levels {
+            // If the index is odd, the parent already exists and needs to be updated
+            // If the index is even, we're creating a new parent
+            if index % 2 != 0 {
+                new_hash = hash(self.hashes[i - 1][index - 1], Some(new_hash));
+                index /= 2;
+                self.hashes[i][index] = new_hash;
+            } else {
+                new_hash = hash(new_hash, Some(new_hash));
+                index /= 2;
+                self.hashes[i].push(new_hash);
+            };
+        }
+
+        self.count - 1
+    }
+
+    // Returns the root hash of the tree
+    pub fn root(&self) -> Option<&u64> {
+        self.hashes[self.levels - 1].first()
+    }
+
+    // A Merkle Tree can generate a proof that it contains an element.
+    // A proof consists of the other hashes that are required to compare with the
+    // root of the tree (assuming the user knows the hash of their data)
+    // Because we have already generated it, it's fairly easy to make
+    // In case of a tree too small or index out of range, returns an empty proof
+    // TODO: for more than an mvp, consider changing this to a result, and send
+    // an OutOfRangeIndex error
```

Comment:
I think this is clear too:

```suggestion
    // TODO: Consider changing this to a result, and send an OutOfRangeIndex error
```